token-swap/program/src/curve/stable.rs

```diff
@@ -175,16 +173,27 @@ impl CurveCalculator for StableCurve {
         swap_token_b_amount: u128,
         round_direction: RoundDirection,
     ) -> Option<TradingTokenResult> {
-        pool_tokens_to_trading_tokens(
-            pool_tokens,
-            pool_token_supply,
-            swap_token_a_amount,
-            swap_token_b_amount,
-            round_direction,
-        )
+        let pool_token_amount = PreciseNumber::new(pool_tokens)?;
+        let pool_token_total_supply = PreciseNumber::new(pool_token_supply)?;
+        let pool_ratio = pool_token_amount.checked_div(&pool_token_total_supply)?;
+        let token_a_amount = PreciseNumber::new(swap_token_a_amount)?;
+        let token_a_value = token_a_amount.checked_mul(&pool_ratio)?;
+        let token_b_amount = PreciseNumber::new(swap_token_b_amount)?;
+        let token_b_value = token_b_amount.checked_mul(&pool_ratio)?;
+        match round_direction {
+            RoundDirection::Floor => Some(TradingTokenResult {
+                token_a_amount: token_a_value.floor()?.to_imprecise()?,
+                token_b_amount: token_b_value.floor()?.to_imprecise()?,
+            }),
+            RoundDirection::Ceiling => Some(TradingTokenResult {
+                token_a_amount: token_a_value.ceiling()?.to_imprecise()?,
+                token_b_amount: token_b_value.ceiling()?.to_imprecise()?,
+            }),
+        }
```

Comment:
This function needs to be inverse of `trading_tokens_to_pool_tokens`, otherwise you'll create an arbitrage opportunity within your own curve.  As it stands, it doesn't seem like that would be the case, since this is the same as the constant product curve.

Reply:
Hey! My understanding is that this is the inverse of `trading_tokens_to_pool_tokens` for the sable curve, even though its logically the same as the constant product curve. This is true is because `pool_tokens_to_trading_tokens` can be calculated without knowledge of the curve (as long as the ratio between pool tokens to pool tokens supply is equal to trading tokens to trading tokens supply), whereas `trading_tokens_to_pool_tokens` is dependent on the curve. We will write tests cases to verify this.

Reply:
They should both be dependent on the curve, since the curve defines the value of tokens contained.  You should try implementing this test https://github.com/solana-labs/solana-program-library/blob/master/token-swap/program/src/curve/calculator.rs#L183 for the stable curve and see what happens!